src/Simulation/CSharpGeneration/EntryPoint.fs

```diff
@@ -119,65 +97,169 @@ let private callableTypeNames context (callable : QsCallable) =
+    let defaultExecutionTarget =
+        context.assemblyConstants.TryGetValue AssemblyConstants.ExecutionTarget
+        |> (fun (_, value) -> if value = null then "" else value)
+        |> literal
+    [
+        namedArg "simulatorOptionAliases" simulatorOptionAliases
+        namedArg "quantumSimulatorName" <| literal AssemblyConstants.QuantumSimulator
+        namedArg "toffoliSimulatorName" <| literal AssemblyConstants.ToffoliSimulator
+        namedArg "resourcesEstimatorName" <| literal AssemblyConstants.ResourcesEstimator
+        namedArg "defaultSimulatorName" <| literal defaultSimulator
+        namedArg "defaultExecutionTarget" <| defaultExecutionTarget
+        namedArg "createDefaultCustomSimulator" <| customSimulatorFactory defaultSimulator
+    ]
+    |> SyntaxFactory.SeparatedList
+    |> SyntaxFactory.ArgumentList
+    |> newDriverSettings.WithArgumentList
+    :> ExpressionSyntax
+
+/// The main method for the standalone executable.
+let private mainMethod context (entryPoints : seq<QsCallable>) =
+
+    let entryPointArrayMembers =
+        [
+            for ep in entryPoints do
+                let name = generateEntryPointClassName ep
+                ``new`` (``type`` (name.ToString())) ``(`` [] ``)``
+        ]
+
+    let entryPointArray =
+        ``new array`` (Some (driverNamespace + ".IEntryPoint")) entryPointArrayMembers
+
+    let driver = ``new`` (``type`` (driverNamespace + ".Driver")) ``(`` [driverSettings context; entryPointArray] ``)``
+    let commandLineArgsName = "args"
+    arrow_method "System.Threading.Tasks.Task<int>" "Main" ``<<`` [] ``>>``
+        ``(`` [param commandLineArgsName ``of`` (``type`` "string[]")] ``)``
+        [``private``; ``static``; async]
+        (Some (``=>`` (await (driver <.> (ident "Run", [ident commandLineArgsName])))))
+
+/// Generates a namespace for the main function
+let mainNamespace context (entryPoints : seq<QsCallable>) =
+    let mainClass =
+        ``class`` entryPointClassName ``<<`` [] ``>>``
+            ``:`` None ``,`` []
+            [``internal``]
             ``{``
-                (Seq.map using SimulationCode.autoNamespaces)
-                [entryPointClass context entryPoint]
+                [mainMethod context entryPoints]
             ``}``
-    ``compilation unit`` [] [] [ns]
+
+    ``namespace`` entryPointClassName
+        ``{``
+            []
+            [mainClass]
+        ``}``
+
+/// Generates C# source code for a standalone executable that runs the Q# entry point.
+let generateSource context (entryPoints : seq<QsCallable>) (mainNamespace : NamespaceDeclarationSyntax option) =
+    let entryPointNamespaces = entryPoints |> Seq.groupBy (fun ep -> ep.FullName.Namespace)
+
+    let epNamespaces = [for ns, eps in entryPointNamespaces -> entryPointNamespace context ns eps]
+
+    let namespaces =
+        match mainNamespace with
+        | Some mainNamespace -> (mainNamespace :> MemberDeclarationSyntax) :: epNamespaces
```

Comment:
The parameter's type is `NamespaceDeclarationSyntax` to enforce that what is passed in should be a namespace, instead of any of the other possible kinds of `MemberDeclarationSyntax` types. However, the `compilation unit` function expects a list of `MemberDeclarationSyntax` objects for the namespaces. This is why `entryPointNamespace` returns a list of `MemberDeclarationSyntax` objects instead of `NamespaceDeclarationSyntax` objects, even though that's what they really are. That's why this casting occurs here.

Reply:
Alternatively, I could return a list of `NamespaceDeclarationSyntax` objects from `entryPointNamespace` (not doing the cast at the end of that function) and create a list of `NamespaceDeclarationSyntax` objects. Then do a typecast across the list when I go to pass it to the `compilation unit` function. That would unify where the casting happens. Yah I think I'll do that, and that will make things a bit clearer.

Reply:
done